rust/stub/src/pe_loader.rs

```diff
@@ -0,0 +1,91 @@
+use goblin::pe::PE;
+use uefi::{
+    prelude::BootServices,
+    table::{
+        boot::{AllocateType, MemoryType},
+        Boot, SystemTable,
+    },
+    Handle, Status,
+};
+
+pub struct Image {
+    pub base: *const u8,
+    pub length: usize,
+    pub entry: extern "efiapi" fn(Handle, SystemTable<Boot>) -> Status,
+}
+
+#[cfg(target_arch = "x86_64")]
+fn flush_instruction_cache(_start: *const u8, _length: usize) {
+    // x86_64 mandates coherent instruction cache
+}
+
+/// Loads and relocates a trusted PE file, returning a structure containing its allocated virtual memory range and entry point.
+pub unsafe fn load(boot_services: &BootServices, file_data: &[u8]) -> uefi::Result<Image> {
+    let pe = PE::parse(file_data).expect("Malformed PE file");
+
+    // Allocate all memory the image will need in virtual memory.
+    // We follow shim here and allocate as EfiLoaderCode.
+    let mut pos = 0;
+    for section in &pe.sections {
+        assert!(section.virtual_address >= pos);
+        pos = section.virtual_address + section.virtual_size;
```

Comment:
This code find the end of the last section? If so you could say:

`pe.sections.iter().map(|s| s.virtual_address + s.virtual_size).max_element()`

This would also remove the limitations that the sections need to be ordered in the PE image.

When your at it, it would help readability to select a better name than `pos`. `virt_end`?